### ISSUE TITLE
enter! macro

### ISSUE THREAD

**Issue by yaahallo** [tokio-rs/tracing#158]
Right now there is some mandatory boilerplate whenever using span!.

```rust
let span = span!(...);
let _something = span.enter();
```

It cannot be written as

```rust
span!(...).enter(); // temporary value freed at end of statement

// or

let span = span!(...);
let _ = span.enter(); // silently fails to do anything, guard is dropped immediately here
```

In order for it to work you must bind the span to a temporary and you must bind the guard to another ignored temporary and that cannot be `_` because of rust drop semantics.

# The Idea

Here enters the enter! macro.

```rust
enter!(span!(...));
```

which should just expand to the first example.

**Comment by hawkw** [comment]
I like the idea of having an `enter!` macro that takes an expression expanding to a `Span`, rather than taking the arguments and forwarding them to the `span!` macro, as it means we don't need to have `enter_trace!`, `enter_debug!`, and so on. 

Should be quite easy to add a macro like this:
```rust
macro_rules! enter {
    ($span:expr) => {
        let span = $span;
        let _guard = span.enter();
    }
}
```

**Comment by singpolyma-shopify** [comment]
Will this expand to inline statements or to a block?  (If a block, it would still drop the guard right away, yes?)

My only concern here would be "polluting" the local namespace with these two extra variables that the programmer didn't create on purpose.

**Comment by yaahallo** [comment]
Aren't rust macros hygenic? I thought that their variables were inaccessible from outside of the macros scope.

**Comment by singpolyma-shopify** [comment]
@yaahallo right, that was the first part of my question.  If this is rendered as a scope then the guard would be dropped right away.

**Comment by yaahallo** [comment]
Ooh, my understanding of declarative macro hygiene is that tracks which identifiers came from inside a macro and disallows using those identifiers outside the macros, it does not logically add `{ }` around the output of the macro to prevent identifiers from escaping.

https://rust-lang.github.io/rustc-guide/macro-expansion.html#hygiene

The best part is the struct in the compiler that tracks the context about where these identifiers are defined is called `Span`.

**Comment by hawkw** [comment]
> Will this expand to inline statements or to a block? (If a block, it would still drop the guard right away, yes?)

In order for this to provide the expected behaviour, this would _have_ to expand to inline statements.
I don't remember OTTOMH whether macro hygiene would permit this or not; if it does not, I think  tokio-rs/tokio#1162 might be the only viable solution for #192.

**Comment by yaahallo** [comment]
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e77fe2ec7e262deff89b78551c7217ce

It works fine in all the ways we need.

**Comment by hawkw** [comment]
> https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e77fe2ec7e262deff89b78551c7217ce
> 
> It works fine in all the ways we need.

In that case...
![PRs_welcome](https://user-images.githubusercontent.com/2796466/61567404-396d5a80-aa34-11e9-9a3f-106a2fb74189.jpg)

**Comment by hawkw** [comment]
@yaahallo is this something you're interested in implementing?

**Comment by yaahallo** [comment]
```rust
/// Macro for explicitly entering a span.
///
/// ```rust
/// use tracing_macros::enter;
/// use tracing::{span, Level};
/// let yak = 2;
/// {
///     enter!(span!(Level::ERROR, "shave", yak = yak));
///     // ... do things within the span
/// }
/// // ... do things after closing and exiting the span
/// ```
#[macro_export]
macro_rules! enter {
    ( $cmd:expr ) => {
        let span = $cmd;
        let _enter = span.enter();
    };
}
```

**Comment by hawkw** [comment]
As a thought, we _might_ want to consider having the `enter!` macro just take a set of arguments that are passed to the `span!` macro, since the main use-case AFAICT is for constructing a new span that should be immediately entered, and it's a _little_ awkward to have to say `enter!(span!(...))`.

**Comment by yaahallo** [comment]
I would prefer this ^ over the version I pasted. I just don't know about calling it `enter` in that case, because ultimately its a span. No idea what I'd call it instead tho.

**Comment by pickfire** [comment]
Is there a reason why `enter` is required? Why not `span` just handle that? Maybe is it related to zero-cost?

```rust
let span = span!(...);
...
// ownership lost, guard automatically dropped.
```

**Comment by tobz** [comment]
@pickfire It allows a single span to be entered multiple times.  As an example, consider a future which is wrapped in a span.  Logically, it must be represented by a single span, but "entering" (and exiting) allows the span to also track when the future is polled.

Constructing a span every time the future was polled wouldn't be the same, semantically.

**Comment by pickfire** [comment]
Yeah, for multiple times it is understandable. But does not it look weird with
```rust
let _ = span.enter();
```
Yes, this patch helps to avoid that. I was wondering if there are any alternatives for single span to be entered only once?

**Comment by hawkw** [comment]
>> Is there a reason why `enter` is required? Why not `span` just handle that? Maybe is it related to zero-cost?
>> 
>> ```rust
>> let span = span!(...);
>> ...
>> // ownership lost, guard automatically dropped.
>> ```
>
> @pickfire It allows a single span to be entered multiple times. As an example, consider a future which is wrapped in a span. Logically, it must be represented by a single span, but "entering" (and exiting) allows the span to also track when the future is polled.
> 
> Constructing a span every time the future was polled wouldn't be the same, semantically.

To add on to this: there are also a number of cases where spans need to be created but not entered, and then entered later in another location. 

For example, consider the case of the `tracing-futures` [`Instrument` combinator](https://docs.rs/tracing-futures/0.0.1-alpha.1/tracing_futures/trait.Instrument.html#method.instrument). We might want to write code like this:

```rust
let my_future = function_returning_a_future()
   .and_then(|x| ...)
   .map(|y| ...)
   .instrument(trace_span!("my future", ...));

// do other stuff...
info!("doing things before spawning my future");
// ...

tokio::spawn(my_future);

// ...
```
In this case, we'd expect the event "doing things before spawning my future" to occur outside the span "my future".  The span won't be entered until we actually spawn the future and it begins executing, so we get the behavior we would expect in this case. This is also important if we want to use the span for profiling the amount of time the future is executing for — if we entered it immediately, rather than only entering it when the future begins executing, that timing would be incorrect.

> Yes, this patch helps to avoid that. I was wondering if there are any other alternatives for single span to be entered only once?

As an alternative to the `enter!` macro proposed here (that might provide nicer ergonomics?), we could also make a change more like the one proposed in #192 (and implemented in the PR tokio-rs/tokio#1162, which wasn't actually merged), which would allow us to write code like this:
```rust
let _enter = span!(...).enter_owned();
```

We can't currently do this with the `.enter()` method, since it _borrows_ the span, and the span needs to be assigned to its own let binding before it can be entered. However, an `enter_owned`/`enter_unscoped` method as proposed in #192 would take ownership of the span, so it doesn't need its own `let` to keep it alive in order to be borrowed. This would make the use case where a span is entered once in a scope a little less verbose without requiring a new macro.

I'd be happy to revive PR tokio-rs/tokio#1162 if people think this would be a better approach?

**Comment by yaahc** [comment]
Does it have to be named still? as in it cant be `let _ = span!(...).enter_owned()`.

If so then I'm against this version personally because I think it's still gonna be a point of friction for new people. Otherwise I definitely like this more than the `enter!(span!(...))` version, for the same reason used to justify postfix await.

**Comment by hawkw** [comment]
> Does it have to be named still? as in it cant be `let _ = span!(...).enter_owned()`.

It will always have to be named, as `let _name` and `let _` have different semantics: `let _` immediately drops the value assigned to it, while `let _name` creates an actual binding that keeps the value alive until the end of the scope as normal. See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a69b20ae651372cb7cc4205c4f8fa7ec) for a demo. 

> If so then I'm against this version personally because I think it's still gonna be a point of friction for new people. 

I agree that this has the potential to be confusing, but it's worth noting that this is generally the case for any guard that is used to indicate a scope but is otherwise unused; the guard will always need to be assigned to a named `let` binding. So, new `tracing` users who do have prior Rust experience may have seen similar code elsewhere, but it is definitely still surprising to new _Rust_ users.

With that said, we could always provide both an `enter!` macro _and_ an `enter_owned` method, since it's worth noting that there is one _other_ potential use case for an `enter_owned` method that the `enter!` macro wouldn't satisfy. In cases where we want a span to remain entered for the entire lifetime of a heap-allocated object, we'd want an `enter_owned` method. The owned guard returned by an `enter_owned`-type method could be used as a field of a struct without making the struct generic over a lifetime. This is admittedly a niche use-case though (and I haven't seen anyone actually trying to do it), in general, it makes more sense for a heap-allocated object to hold a `Span` and enter that span in a scope whenever methods on that object are called. Also, having both has the downside of adding more ways of doing similar things, which I think tends to be confusing...

**Comment by yaahc** [comment]
I'll leave it up to you.

**Comment by pickfire** [comment]
> > Does it have to be named still? as in it cant be let _ = span!(...).enter_owned().
>
> It will always have to be named, as let _name and let _ have different semantics: let _ immediately drops the value assigned to it, while let _name creates an actual binding that keeps the value alive until the end of the scope as normal. See this playground for a demo.

Looks like a potential footgun to me. Some beginners might just do one of the following but no errors are being displayed if I am correct. (I did not verify these but these are what comes to my mind)

```rust
let _ = span!();
let _name = span!().enter_owned();
span!().enter_owned();
enter!(span!());
```

**Comment by hawkw** [comment]
> > > Does it have to be named still? as in it cant be let _ = span!(...).enter_owned().
> > 
> > 
> > It will always have to be named, as let _name and let _ have different semantics: let _ immediately drops the value assigned to it, while let _name creates an actual binding that keeps the value alive until the end of the scope as normal. See this playground for a demo.
> 
> Looks like a potential footgun to me. Some beginners might just do one of the following but no errors are being displayed if I am correct. (I did not verify these but these are what comes to my mind)
> 
> ```rust
> let _ = span!();
> let _name = span!().enter_owned();
> span!().enter_owned();
> enter!(span!());
> ```

Not all of these would cause incorrect behavior --- only the first and third would result in the span being immediately exited. The third example you gave (`span!().enter_owned()`) would be caught by a `must_use` attribute on the guard type, so the user would be warned that it could cause incorrect behavior. 

In the first case (`let _ = span!()`), the user is explicitly choosing to immediately discard the span, ignoring any `must_use` attributes; this is the behavior that `let _` has in all Rust code. For example, `let _` is often used to explicitly mark that any potential error returned by a function returning `Result` is being intentionally ignored. 

> Are there any ways to guard them from writing these? Maybe?
> 
> ```rust
> enter!(span!(...), || do_something());
> enter!(..., || do_something());
> ```

This seems similar to the [`Span::in_scope`](https://docs.rs/tracing/0.1.9/tracing/struct.Span.html#method.in_scope) method.

**Comment by pickfire** [comment]
Nice explanations.

>> let _ = span!();
>
> In the first case (let _ = span!()), the user is explicitly choosing to immediately discard the span, ignoring any must_use attributes; this is the behavior that let _ has in all Rust code. For example, let _ is often used to explicitly mark that any potential error returned by a function returning Result is being intentionally ignored.

Okay, so that is the only possible case not guarded by rust. How does the library prevents the users from doing that?

This is solvable with documentation but leaving a potential foot gun to the users by assuming everyone read the docs (or if they did not missed it) seems bad.

**Comment by yaahc** [comment]
For posterity, in case anyone else runs into this itch, I've managed to solve the issue locally by adding the following helper macro to my crates.

```rust
/// Helper macro to enter spans conveniently
#[macro_export]
macro_rules! spanned {
    ($local:ident) => {
        let span = span!(Level::INFO, stringify!($local), $local);
        let _guard = span.enter();
    };
    (?$local:ident) => {
        let span = span!(Level::INFO, stringify!($local), ?$local);
        let _guard = span.enter();
    };
    ($level:expr, $local:ident) => {
        let span = span!($level, stringify!($local), $local);
        let _guard = span.enter();
    };
    ($level:expr, ?$local:ident) => {
        let span = span!($level, stringify!($local), ?$local);
        let _guard = span.enter();
    };
    ( $( $span:tt )* ) => {
        let span = span!($($span)*);
        let _guard = span.enter();
    };
}
```

**Comment by hawkw** [comment]
@yaahc fwiw, I am definitely open to considering upstreaming a macro like that, I'm just a _little_ concerned about the maintenance burden of macros...if it's something a lot of people find useful, we can definitely bring it into `tracing`.

**Comment by yaahc** [comment]
would going from this

```rust
( $( $span:tt )* ) => {
        let span = span!($($span)*);
        let _guard = span.enter();
    };
```

to this

```rust
( $( $span:tt )* ) => {
        let span = span!($($span)*);
        let _guard = span.enter();
        span
    };
```

solve the latter problem?
